keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,847 @@
+	// "UnlimitedSwap": workloads can use unlimited swap, up to the allocatable limit.
+	// "LimitedSwap": workloads can use up to this limit of swap.
+	SwapBehavior string
+
+	LimitedSwap *LimitedSwapConfiguration
+}
+
+type LimitedSwapConfiguration struct {
+	PerWorkloadMemorySwapLimit resource.Quantity
+}
+```
+
+We want to expose common swap configurations based on the [Docker] and open
+container specification for the `--memory-swap` flag. Thus, the
+`MemorySwapConfiguration.SwapBehavior` setting will have the following effects:
+
+* If `SwapBehavior` is not set or set to `"NoSwap"`, containers do not have
+  access to swap. This value effectively prevents a container from using swap,
+  even if it is enabled on a system.
+* If `SwapBehavior` is set to `"UnlimitedSwap"`, the container is allowed to
+  use unlimited swap, up to the maximum amount available on the host system.
+* If `SwapBehavior` is set to `"LimitedSwap"`, then the `LimitedSwap`
+  configuration must also be set. `LimitedSwap.PerWorkloadMemorySwapLimit`
+  represents the system-wide maximum limit for swap usage of a container. Note
+  that this limit applies to individual containers, and not at the pod-level,
+  in order to be set via the CRI rather than e.g. a [pod cgroup limit].
+
+[docker]: https://docs.docker.com/config/containers/resource_constraints/#--memory-swap-details
+[pod cgroup limit]: #restrict-swap-usage-at-the-cgroup-level
+
+#### CRI Changes
+
+The CRI requires a corresponding change in order to allow the kubelet to set
+swap usage in container runtimes.  We will introduce a parameter
+`memory_swap_limit_in_bytes` to the CRI API (found in
+[k8s.io/cri-api/pkg/apis/runtime/v1/api.proto]):
+
+[k8s.io/cri-api/pkg/apis/runtime/v1/api.proto]: https://github.com/kubernetes/kubernetes/blob/6baad0a1d45435ff5844061aebab624c89d698f8/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L563-L580
+
+```go
+// LinuxContainerResources specifies Linux specific configuration for
+// resources.
+message LinuxContainerResources {
+...
+    // Memory + swap limit in bytes. Default: 0 (not specified).
+    int64 memory_swap_limit_in_bytes = 9;
+...
+}
+```
+
+### Test Plan
+
+For alpha:
+
+- Swap scenarios are enabled in test-infra for at least two Linux
+  distributions. e2e suites will be run against them.
+  - Container runtimes must be bumped in CI to use the new CRI.
+- Data should be gathered from a number of use cases to guide beta graduation
+  and further development efforts.
+  - Focus should be on supported user stories as listed above.
+
+Once this data is available, additional test plans should be added for the next
+phase of graduation.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Kubelet can be started with swap enabled.
+- KubeletConfig allows CRI to be configured with a percentage of swap available
+  to workloads. This will default to 0.
+- e2e test jobs are configured for Linux systems with swap enabled.
+
+#### Beta
+
+(Tentative.)
```

Comment:
- handle usage of swap during container restart boundaries for writes to tmpfs (which may require pod cgroup change beyond what container runtime will do at container cgroup boundary).
- enable ability to reserve swap for kube/system-reserved from what kubelet detects on host
- better understand relationship of swap with memory qos in cgroup v2 (particularly memory.high usage)